src/System.Management.Automation/engine/Subsystem/SubsystemInfo.cs

```diff
@@ -0,0 +1,349 @@
+            }
+
+            /// <summary>
+            /// Gets the unique identifier for a subsystem implementation.
+            /// </summary>
+            public Guid Id { get; }
+
+            /// <summary>
+            /// Gets the kind of subsystem.
+            /// </summary>
+            public SubsystemKind Kind { get; }
+
+            /// <summary>
+            /// Gets the name of a subsystem implementation.
+            /// </summary>
+            public string Name { get; }
+
+            /// <summary>
+            /// Gets the description of a subsystem implementation.
+            /// </summary>
+            public string Description { get; }
+
+            /// <summary>
+            /// Gets the implementation type.
+            /// </summary>
+            public Type ImplementationType { get; }
+        }
+
+        #endregion
+    }
+
+    internal sealed class SubsystemInfoImpl<TConcreteSubsystem> : SubsystemInfo
+        where TConcreteSubsystem : class, ISubsystem
+    {
+        private ReadOnlyCollection<TConcreteSubsystem> _registeredImpls;
+
+        internal SubsystemInfoImpl(SubsystemKind kind)
+            : base(kind, typeof(TConcreteSubsystem))
+        {
+            _registeredImpls = Utils.EmptyReadOnlyCollection<TConcreteSubsystem>();
+        }
+
+        /// <summary>
+        /// The 'add' and 'remove' operations are implemented in a way to optimize the 'reading' operation,
+        /// so that reading is lock-free and allocation-free, at the cost of O(n) copy in 'add' and 'remove'
+        /// ('n' is the number of registered implementations).
+        /// </summary>
+        /// <remarks>
+        /// In the subsystem scenario, registration operations will be minimum, and in most cases, the registered
+        /// implementation will never be unregistered, so optimization for reading is more important.
+        /// </remarks>
+        private protected override void AddImplementation(ISubsystem rawImpl)
+        {
+            lock (_syncObj)
+            {
+                var impl = (TConcreteSubsystem)rawImpl;
+
+                if (_registeredImpls.Count == 0)
+                {
+                    _registeredImpls = new ReadOnlyCollection<TConcreteSubsystem>(new[] { impl });
+                    _cachedImplInfos = new ReadOnlyCollection<ImplementationInfo>(new[] { new ImplementationInfo(impl) });
+                    return;
+                }
+
+                if (!AllowMultipleRegistration)
+                {
+                    throw new InvalidOperationException(
+                        StringUtil.Format(
+                            SubsystemStrings.MultipleRegistrationNotAllowed,
+                            Kind.ToString()));
+                }
+
+                bool targetExists = false;
+                foreach (TConcreteSubsystem item in _registeredImpls)
+                {
+                    if (item.Id == impl.Id)
+                    {
+                        targetExists = true;
```

Comment:
We could exclude the `targetExists` and move `throw` in the line.

Reply:
Updated.